### DISCUSSION TITLE
ACL Built in Database lost - When updating EMQX (CE 5.7.0/1) configuration (via CR)

### DISCUSSION THREAD

**Discussion by ennc0d3** [emqx/emqx#13385] [General]
I am trying to understand the reason behind the loss of ACL built_in_database entries, which appears to immediately happen when I update the EMQX CR.

I am using 5.7.1 (CE) and all the authorized entries in ACL (built_in_database) are lost when I update for example update config.data section, for example the log level. It is bit strange and there is not enough logs suggesting what happened. The number of ACL entries becomes zero after the CR update.

This occurs only when I update the config via the CR and not when i update via the dashboard. I am using the EMQX Operator (2.2.23) with EMQX CE 5.7.1

```
apiVersion: apps.emqx.io/v2beta1
kind: EMQX
metadata:
  name: emqx
  namespace: emqx
spec:
  image: "emqx:5.7.1"
  revisionHistoryLimit: 3
    initialDelaySeconds: 10
    type: Recreate        
  # Configuration Reference: https://www.emqx.io/docs/en/v5.6.1/hocon/
  config:
    data: |
      dashboard {
        listeners.http.bind = 18083
      }
      log {
        console_handler {
          level = "debug"
          enable = true
          formatter = "text"
        }
      }
....
```

Where should I look and do you have some reasons for this behaviour? Thanks

**Comment by zmstone** [comment]
Hi @ennc0d3 
If the pod is restarted, the database will be gone unless you mount a volume for the data dir. (`/opt/emqx/data`).

btw, just FYI.
the log config has a deprecated config key `console_handler`, it should be just `console` now.
`console_hanlder` happens to be the the current type name and an alias for the field name.

![image](https://github.com/emqx/emqx/assets/164324/ff5c7515-95b5-4653-9450-fe2ba15b38e0)

**Reply by ennc0d3** [reply]
The POD mounts a volume for /opt/emqx/data, the rest of the data is still there, what get wiped is the AuthZ DB data. There is no pod restart in this case, the change is reflected in the configuration.

The log field is only an example, If i change any of the MQTT settings, example the max_inflight value, the AuthZ DB disappears too. 

Thanks for the picture, It helps. BTW, the console_handler is an alias to console, so I believe it shouldn't be issue.

**Reply by zmstone** [reply]
databas is in `data/mnesia/{emqx@node.host.name}`, if the container gets a new name (or IP if no hostname/DNS), its get a fresh database.

**Reply by ennc0d3** [reply]
In this case there is no POD changes and I can see the Authn entries persisted in Db, but not AuthZ entries which is something I am able to figure why. 
 
For the second point,  I would like to know how we can have stable identifiers? Because I would like the POD to re-use the PVC, thereby the retained data.

**Reply by ennc0d3** [reply]
> For the second point, I would like to know how we can have stable identifiers? Because I would like the POD to re-use the PVC, thereby the retained data.

I am using EMQX operator, I don't see a field that I can use to set stable name. As I understand when the name changes, we get a new db storage. But, when I upgrade the EMQX cluster, I could still see the old data though the name of the core node has changed (example, K8s attaches the hash suffix).

Trying to understand the mechanism, Is it because of the data is imported automatically?

**Reply by zmstone** [reply]
I believe operator should assign static (DNS based) names for the nodes. @Rory-Z  is that correct ?

If not, continue reading:
If the nodes are upgraded one after another (rolling), the newly created node will copy configs, cert files, plugins, and database from the "config leader" node.

**Reply by ennc0d3** [reply]
When we upgrade the hash that is used in the node name changes, thereby we have a new PVC. (New stateful set with a different hash). So if I understand your point, there is a logic to copy the data from the OLD "config leader" node. Correct me if am wrong. Also can you point me to the document or relevant code to confirm the behaviour.

**Comment by zmstone** [comment]
> Thanks for the picture, It helps. BTW, the console_handler is an alias to console, so I believe it shouldn't be issue.

Yes, that's what I wrote (that it's an alias).
But it's not documented

**Reply by ennc0d3** [reply]
Adding the EMQX CR, if this helps to understand

```cat /tmp/emqx.yaml                                                                                                                                                                                                                                        
apiVersion: apps.emqx.io/v2beta1
kind: EMQX
metadata:
  name: emqx
  namespace: cloud
spec:
  bootstrapAPIKeys:
  - key: emqx_basic_token
    secret: XXX
  config:
    data: |
      dashboard {
        listeners.http.bind = 18083
      }
      log {
        console {
          level = "debug"
          enable = true
          formatter = "text"
        }
      }
      listeners.tcp.default.enable = false
      listeners.ssl.default.enable = true
      listeners.ssl.default.max_conn_rate = "1024/s"
      listeners.ssl.default.messages_rate = "256/s"
      listeners.ssl.default.bytes_rate = "1MB/s"
      listeners.ssl.default.ssl_options.certfile = "/mnt/certs/tls.crt"
      listeners.ssl.default.ssl_options.keyfile = "/mnt/certs/tls.key"
      listeners.ws.default.enable = false
      listeners.wss.default.enable = false
      retainer.backend.storage_type = disc
      mqtt.max_mqueue_len = 300000
      mqtt.max_inflight = 256
      sysmon.vm.long_schedule = disabled
      telemetry.enable = false
      authentication = [
        {
          use_jwks = true
          endpoint = "http://x.y.z.0"
          pool_size = 8
          refresh_interval = 300
          ssl = {
            enable = false
          }
          mechanism = jwt
          acl_claim_name = acl
          verify_claims = {
            aud = "xxx"
            iss = "${clientid}"
          }
          from = password
          enable = true
        },
        {
          mechanism = password_based
          backend = built_in_database
          user_id_type = username
          password_hash_algorithm = {
            name = sha256
            salt_position = prefix
          }
          enable = true
        },
      ]
      authorization {
        no_match = deny
        deny_action = disconnect
        sources = [
          {
            type = built_in_database
            enable = true
          },
          {
            type = file
            enable = true
            # This file is immutable to EMQX.
            # Once new rules are created from dashboard UI or HTTP API,
            # the file 'data/authz/acl.conf' is used instead of this one
            path = "etc/acl.conf"
          }
        ]
      }
    mode: Merge
  coreTemplate:
    metadata:
      name: emqx-core
    spec:
      extraVolumeMounts:
      - mountPath: /mnt/certs
        name: emqx-tls-secrets-vol
      - mountPath: /opt/emqx/etc/acl.conf
        name: emqx-acl-config-vol
        subPath: acl.conf
      extraVolumes:
      - name: emqx-tls-secrets-vol
        secret:
          secretName: emqx-tls-secrets
      - configMap:
          name: emqx-acl-configmap
        name: emqx-acl-config-vol
      replicas: 2
      volumeClaimTemplates:
        accessModes:
        - ReadWriteOnce
        resources:
          requests:
            storage: 10Gi
        storageClassName: standard
  image: emqx:5.7.1
  replicantTemplate:
    metadata:
      name: emqx-replicant
      extraVolumeMounts:
      - mountPath: /mnt/certs
        name: emqx-tls-secrets-vol
      - mountPath: /opt/emqx/etc/acl.conf
        name: emqx-acl-config-vol
        subPath: acl.conf
      extraVolumes:
      - name: emqx-tls-secrets-vol
        secret:
          secretName: emqx-tls-secrets
      - configMap:
          name: emqx-acl-configmap
        name: emqx-acl-config-vol
      replicas: 1
  revisionHistoryLimit: 3
  updateStrategy:
    initialDelaySeconds: 10
    type: Recreate
    ```

**Comment by zmstone** [comment]
> In this case there is no POD changes and I can see the Authn entries persisted in Db, but not AuthZ entries which is something I am able to figure why.

I managed to reproduce it in below steps without k8s.

1. Put this in a file `path/to/authz.hocon`
```
authorization {
  no_match = deny
  deny_action = disconnect
  sources = [
    {
      type = built_in_database
      enable = true
    },
    {
      type = file
      enable = true
      path = "${EMQX_ETC_DIR}/acl.conf"
    }
  ]
}
```
2. Run `emqx ctl conf load path/to/authz.hocon`
3. Go to dashboard, add a new rule in built-in database
4. Run command `emqx ctl conf load path/to/authz.hocon` again
5. The database table is cleared.

Will fix it in 5.7.2

**Reply by zhongwencool** [reply]
https://github.com/emqx/emqx/pull/13398

**Reply by ennc0d3** [reply]
Do you have a release date? Thanks

**Reply by zmstone** [reply]
Release date is set on Jul 25, 2024.
With a risk of delay, but should not be long.
